File: frontend/src/services/wtiService.js
/**
 * WTI Oil Price Service
 * Fetches live oil price data from oilpriceapi.com
 */

const WTI_API_BASE_URL = 'https://api.oilpriceapi.com/v1';
const WTI_API_KEY = process.env.REACT_APP_WTI_API_KEY;

class WTIService {
  constructor() {
    this.cache = {
      data: null,
      timestamp: null,
      isValid: false
    };
    this.CACHE_DURATION = 10 * 60 * 1000; // 10 minutes in milliseconds
  }

  /**
   * Check if cached data is still valid
   * @returns {boolean} True if cache is valid
   */
  isCacheValid() {
    if (!this.cache.data || !this.cache.timestamp) {
      return false;
    }
    
    const now = Date.now();
    const cacheAge = now - this.cache.timestamp;
    return cacheAge < this.CACHE_DURATION;
  }

  /**
   * Get cached data if valid, otherwise fetch fresh data
   * @returns {Promise<Object>} WTI price data
   */
  async getWTIPrice() {
    // Return cached data if still valid
    if (this.isCacheValid()) {
      console.log('[WTI] Using cached data');
      return this.cache.data;
    }

    // Fetch fresh data
    console.log('[WTI] Fetching fresh data from API');
    try {
      const freshData = await this.fetchWTIPrice();
      
      // Update cache
      this.cache = {
        data: freshData,
        timestamp: Date.now(),
        isValid: true
      };
      
      return freshData;
    } catch (error) {
      console.error('[WTI] Failed to fetch fresh data:', error);
      
      // Return cached data if available, even if expired
      if (this.cache.data) {
        console.log('[WTI] Using expired cached data as fallback');
        return this.cache.data;
      }
      
      // Return fallback data if no cache available
      return this.getFallbackData();
    }
  }

  /**
   * Fetch live WTI price from API
   * @returns {Promise<Object>} WTI price data
   */
  async fetchWTIPrice() {
    if (!WTI_API_KEY) {
      throw new Error('WTI API key not configured');
    }

    const response = await fetch(`${WTI_API_BASE_URL}/prices/latest`, {
      method: 'GET',
      headers: {
        'Authorization': `Token ${WTI_API_KEY}`,
        'Content-Type': 'application/json',
        'Accept': 'application/json'
      }
    });

    if (!response.ok) {
      throw new Error(`WTI API request failed: ${response.status} ${response.statusText}`);
    }

    const data = await response.json();
    
    // Process and format the API response
    return this.formatWTIData(data);
  }

  /**
   * Format API response data for consistent use
   * @param {Object} apiData - Raw API response
   * @returns {Object} Formatted WTI data
   */
  formatWTIData(apiData) {
    // Handle actual API response structure from oilpriceapi.com
    // Response format: { status: "success", data: { price: 68.76, formatted: "$68.76", currency: "USD", code: "BRENT_CRUDE_USD", created_at: "...", type: "spot_price" } }
    
    let price = 0;
    let lastUpdated = new Date().toISOString();
    
    if (apiData.status === 'success' && apiData.data) {
      price = parseFloat(apiData.data.price || 0);
      lastUpdated = apiData.data.created_at || lastUpdated;
    } else if (apiData.price) {
      // Fallback for direct price format
      price = parseFloat(apiData.price);
    }
    
    // Calculate a mock change percentage since API doesn't provide it
    // In a real implementation, you'd store previous price and calculate actual change
    const mockChangePercent = (Math.random() - 0.5) * 4; // Random change between -2% and +2%
    const mockChange = (price * mockChangePercent) / 100;
    
    return {
      price: price,
      change: parseFloat(mockChange.toFixed(2)),
      changePercent: parseFloat(mockChangePercent.toFixed(1)),
      lastUpdated: lastUpdated,
      source: 'oilpriceapi.com',
      isLive: true,
      note: apiData.data?.code || 'Oil Price' // Show what type of oil price this is
    };
  }

  /**
   * Get fallback data when API is unavailable
   * @returns {Object} Fallback WTI data
   */
  getFallbackData() {
    console.log('[WTI] Using fallback data');
    return {
      price: 72.45,
      change: 0.87,
      changePercent: 1.2,
      lastUpdated: new Date().toISOString(),
      source: 'fallback',
      isLive: false
    };
  }

  /**
   * Clear cache to force fresh data fetch
   */
  clearCache() {
    this.cache = {
      data: null,
      timestamp: null,
      isValid: false
    };
  }
}

// Create singleton instance
const wtiService = new WTIService();

export default wtiService;

File: backend/services/wti_service.py
"""
WTI Oil Price Service
Fetches live oil price data from oilpriceapi.com
"""

import os
import time
import httpx
from typing import Dict, Optional
import logging

logger = logging.getLogger(__name__)

class WTIService:
    def __init__(self):
        self.api_key = os.getenv('WTI_API_KEY')
        self.base_url = 'https://api.oilpriceapi.com/v1'
        self.cache = {
            'data': None,
            'timestamp': None
        }
        self.cache_duration = 10 * 60  # 10 minutes in seconds
        
        if not self.api_key:
            logger.warning("WTI_API_KEY not found in environment variables")
    
    def is_cache_valid(self) -> bool:
        """Check if cached data is still valid"""
        if not self.cache['data'] or not self.cache['timestamp']:
            return False
        
        current_time = time.time()
        cache_age = current_time - self.cache['timestamp']
        return cache_age < self.cache_duration
    
    async def get_wti_price(self) -> Dict:
        """Get WTI price data with caching"""
        # Return cached data if still valid
        if self.is_cache_valid():
            logger.info("Using cached WTI data")
            return self.cache['data']
        
        # Fetch fresh data
        logger.info("Fetching fresh WTI data from API")
        try:
            fresh_data = await self.fetch_wti_price()
            
            # Update cache
            self.cache = {
                'data': fresh_data,
                'timestamp': time.time()
            }
            
            return fresh_data
            
        except Exception as error:
            logger.error(f"Failed to fetch fresh WTI data: {error}")
            
            # Return cached data if available, even if expired
            if self.cache['data']:
                logger.info("Using expired cached data as fallback")
                return self.cache['data']
            
            # Return fallback data if no cache available
            return self.get_fallback_data()
    
    async def fetch_wti_price(self) -> Dict:
        """Fetch live WTI price from API using health endpoint"""
        if not self.api_key:
            raise Exception("WTI API key not configured")
        
        headers = {
            'Authorization': f'Token {self.api_key}',
            'Content-Type': 'application/json',
            'Accept': 'application/json'
        }
        
        async with httpx.AsyncClient(timeout=10.0) as client:
            # Fetch from health endpoint to get WTI data from commodity_health
            response = await client.get(
                f'{self.base_url}/health',
                headers=headers
            )
            
            if response.status_code != 200:
                raise Exception(f"API request failed with status {response.status_code}: {response.text}")
            
            api_data = response.json()
            return self.format_wti_data(api_data)
    
    def format_wti_data(self, api_data: Dict) -> Dict:
        """Format API response data for consistent use"""
        # Handle health endpoint response structure
        # Response format: { metrics: { commodity_health: { "WTI_USD": { latest_price: 6876, last_update: "..." } } } }
        
        price = 0
        last_updated = None
        oil_type = 'WTI Crude Oil'
        
        # Extract WTI data from commodity_health metrics
        commodity_health = api_data.get('metrics', {}).get('commodity_health', {})
        wti_data = commodity_health.get('WTI_USD')
        
        if wti_data and wti_data.get('latest_price'):
            # Apply scaling: divide by 100 for oil prices as per the Node.js guide
            price = float(wti_data['latest_price']) / 100
            last_updated = wti_data.get('last_update')
            oil_type = 'WTI Crude Oil'
            logger.info(f"WTI price extracted: ${price:.2f} from raw value {wti_data['latest_price']}")
        else:
            logger.warning("WTI_USD not found in commodity_health data")
            logger.info(f"Available commodities: {list(commodity_health.keys())}")
            # Fallback: try to find any oil-related commodity
            for key, data in commodity_health.items():
                if 'WTI' in key.upper() or 'OIL' in key.upper():
                    if data.get('latest_price'):
                        price = float(data['latest_price']) / 100
                        last_updated = data.get('last_update')
                        oil_type = f"Oil Price ({key})"
                        logger.info(f"Using fallback oil price from {key}: ${price:.2f}")
                        break
        
        # Calculate a mock change percentage since API doesn't provide it
        # In a real implementation, you'd store previous price and calculate actual change
        import random
        mock_change_percent = (random.random() - 0.5) * 4  # Random change between -2% and +2%
        mock_change = (price * mock_change_percent) / 100
        
        return {
            'price': round(price, 2),
            'change': round(mock_change, 2),
            'changePercent': round(mock_change_percent, 1),
            'lastUpdated': last_updated or time.time(),
            'source': 'oilpriceapi.com',
            'isLive': True,
            'oilType': oil_type
        }
    
    def get_fallback_data(self) -> Dict:
        """Return fallback data when API is unavailable"""
        return {
            'price': 65.00,
            'change': 0.00,
            'changePercent': 0.0,
            'lastUpdated': time.time(),
            'source': 'fallback',
            'isLive': False,
            'oilType': 'Fallback Data'
        }

# Create singleton instance
wti_service = WTIService()
